### ISSUE TITLE
An error occurred:[Errno socket error] [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:581)

### ISSUE THREAD

**Issue by cilahn** [NagiosEnterprises/ncpa#195]
Hello,
i just toyed around a bit with ncpa and followed the Instructions at https://assets.nagios.com/downloads/ncpa/docs/html/. Everything looked fine until i tried to reach the ncpa listener from my nagios host with check_ncpa.py.
As it seems there is something i'm missing in the ssl cert handling and check_ncpa.py quits with the following error:
./check_ncpa.py -H frodo -t <token> --list -s
The stack trace:Traceback (most recent call last):
  File "./check_ncpa.py", line 266, in main
    info_json = get_json(options)
  File "./check_ncpa.py", line 220, in get_json
    filename, _ = urlretrieve(url)
  File "/usr/lib/python2.7/urllib.py", line 98, in urlretrieve
    return opener.retrieve(url, filename, reporthook, data)
  File "/usr/lib/python2.7/urllib.py", line 245, in retrieve
    fp = self.open(url, data)
  File "/usr/lib/python2.7/urllib.py", line 213, in open
    return getattr(self, name)(url)
  File "/usr/lib/python2.7/urllib.py", line 443, in open_https
    h.endheaders(data)
  File "/usr/lib/python2.7/httplib.py", line 997, in endheaders
    self._send_output(message_body)
  File "/usr/lib/python2.7/httplib.py", line 850, in _send_output
    self.send(msg)
  File "/usr/lib/python2.7/httplib.py", line 812, in send
    self.connect()
  File "/usr/lib/python2.7/httplib.py", line 1212, in connect
    server_hostname=server_hostname)
  File "/usr/lib/python2.7/ssl.py", line 350, in wrap_socket
    _context=self)
  File "/usr/lib/python2.7/ssl.py", line 566, in __init__
    self.do_handshake()
  File "/usr/lib/python2.7/ssl.py", line 788, in do_handshake
    self._sslobj.do_handshake()
IOError: [Errno socket error] [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:581)

Parts of my ncpa.cfg:

[listener]
uid = nagios
certificate = adhoc
loglevel = info
ip = 0.0.0.0
gid = nagcmd
logfile = var/ncpa_listener.log
port = 5693
pidfile = var/ncpa_listener.pid
\# Available versions: PROTOCOL SSLv2, SSLv3, TLSv1
ssl_version = TLSv1

I read through the installation instructions but couldn't find a clue for the certificate handling between the plugin check and the listener. Can you help me with that?

Cheers,

Cilahn

**Comment by cilahn** [comment]
I digged deeper into this and found a python "issue" describing the problem.
http://bugs.python.org/issue23052 tells us that the default behavior of checking certificates has been changed in PEP0476 (https://www.python.org/dev/peps/pep-0476/) for python versions  3.4.x, 3.5 and 2.7.X.

I tried out a very rough and unsecure hack to verfify this by adding this to check_ncpa.py:

```
import ssl

try:
    _create_unverified_https_context = ssl._create_unverified_context
except AttributeError:
    # Legacy Python that doesn't verify HTTPS certificates by default
    pass
else:
    # Handle target environment that doesn't support HTTPS verification
    ssl._create_default_https_context = _create_unverified_https_context
```

This fixes the ssl error, but also completly breaks the security of certificates by accepting every cert.

Can this be fixed in future versions with some kind of import mechanism for the ncpa server certs?

Cilahn

**Comment by cilahn** [comment]
one final update. This can be fixed by importing the /usr/local/ncpa/ncpa.crt to the local nagios ssl storage. To do so have  a good read at http://gagravarr.org/writing/openssl-certs/others.shtml#selfsigned-openssl for the basics behind it. Than roughly i did this:
```
1. copy over /usr/local/ncpa/ncpa.crt to /usr/share/ca-certificates/ncpa/<hostname>_ncpa.crt (i run debian)
2. In /etc/ssl ln -s /usr/share/ca-certificates/ncpa/<hostname>_ncpa.crt `openssl x509 -noout -hash -in /usr/share/ca-certificates/ncpa/<hostname>_ncpa.crt`.0
3. Repeat for every ncpa monitored host
```
This imports your server Cert as an CACert because with adhoc you use self signed certificates. Be on your toes and check the fingerprints and directory rights to stay relativly safe. Also the 'hack' mentioned before is not needed.

So please update the online documentation to reflect the PEP proposal.

Cheers,

Cilahn

**Comment by jhomann** [comment]
I've had to do something similar with our commercial products for some integration to work and it is definitely something we could do. I will take a look into this and try and get it into the next version.

**Comment by jhomann** [comment]
This has two problems I see... and I think the solution will be to add a flag to check_ncpa.py that will allow not verifying the SSL cert vs verifying - so that if someone wants the extra security they can. The update seems to be a problem on systems that run > python 2.6/2.7.x up to 2.7.9

**Comment by jhomann** [comment]
This has been fixed with check_ncpa.py 1.0.0 which will be available when 2.0.0 is released.
